### ISSUE TITLE
FTX not updating current rate properly

### ISSUE THREAD

**Issue by Dante2333** [freqtrade/freqtrade#2829]
## Step 1: Have you search for this issue before posting it?

Yes

## Step 2: Describe your environment

  * Operating system: Ubuntu
  * Python Version: 3.7
  * CCXT version: latest
  * Branch:  Develop
  * Last Commit ID: 2020-01-31
  
## Step 3: Describe the problem:

Bot keeps saying he found open orders and wallets synced, no error output but current rate wont update per pair unless bot is restarted from scratch

### Steps to reproduce:

  1. Run FTX on leveraged tokens
  2. Wait a few minutes after a buy signal
  3. Bot will keep saying wallets synced and found open orders
  
### Observed Results:

  * What happened? bots keep syncing and saying wallets synced, and saying found open orders but balance wont update.
  * What did you expect to happen? Balances to update and bot to spam Bot heartbeat

### Relevant code exceptions or logs:

  ```
2020-02-01 04:02:36,469 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=35, pair=ETHBEAR/USD, amount=0.03962750, open_rate=50.47000000, open_since=2020-01-31 16:01:06)
2020-02-01 04:02:36,646 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=36, pair=LEOBULL/USD, amount=0.01200120, open_rate=166.65000000, open_since=2020-01-31 16:01:07)
2020-02-01 04:02:36,852 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=37, pair=LTCBEAR/USD, amount=0.00040917, open_rate=4888.00000000, open_since=2020-01-31 16:01:08)
2020-02-01 04:02:37,824 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=38, pair=XTZBULL/USD, amount=0.00026729, open_rate=7482.50000000, open_since=2020-01-31 16:01:09)
2020-02-01 04:02:37,967 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=39, pair=ALGOBEAR/USD, amount=0.00346320, open_rate=577.50000000, open_since=2020-02-01 00:01:10)
2020-02-01 04:02:38,166 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=40, pair=LTCBULL/USD, amount=0.01457726, open_rate=137.20000000, open_since=2020-02-01 04:01:08)
2020-02-01 04:02:38,585 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:38,794 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:38,983 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:39,185 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:39,583 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:39,783 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:39,994 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:40,185 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:40,389 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:40,586 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:40,791 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:40,989 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:41,187 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:41,392 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:41,587 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:41,795 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:41,984 - freqtrade.wallets - INFO - Wallets synced.
2020-02-01 04:02:42,191 - freqtrade.wallets - INFO - Wallets synced.

  ```

**Comment by xmatthias** [comment]
this log will not suffice to determine any problem.

What happens after this? 
for how long does this go on? 
does it fall back to "ping" eventually?

wallets can be updated from several places - and it always depends on what is happening (have the trades been filled? are the orders still open on the exchange, ...).


also, your descripton and the logs don't really align.
You're talking about "current rate" - but on the other hand about wallets. 

These 2 are not related - so what exactly is the problem other than that you don't like the log-output?

**Comment by Dante2333** [comment]
The output just cycles between these without stopping, never falling back to ping. By current rate I meant on telegram when I query for status, coins will show current rate as an old state that is not true and does not change over time. For example it bought algobear at 355 then its now 375 but in reality so much time has pass and now algobear is 300 but current rate won't update on /status

**Comment by xmatthias** [comment]
then best run the bot using `-v` - so you'll get verbose output.

Also, what's the configuration of `bid_strategy`  and `ask_strategy`?
are you using orderbook? or some other "non-default" configuration?

then please test the methods we use to get the price manually:
initialize a ccxt object for FTX and run the methods `ct.fetch_ticker(pair)` and `ct.fetch_l2_order_book(pair, 10)` and post the output here.


To be honest, these are very severe problems - which increase my belive that FTX will not be supportable by freqtrade in any forseable future.

**Comment by Dante2333** [comment]
I forgot I turned on use orderbook for bid strategy, I do not remember why. I've turned it off and will try again.

**Comment by Dante2333** [comment]
```
{'symbol': 'BCHBEAR/USD', 'timestamp': 1580603211764, 'datetime': '2020-02-02T00:26:51.764Z', 'high': None, 'low': None, 'bid': 324.0, 'bidVolume': None, 'ask': 327.1, 'askVolume': None, 'vwap': None, 'open': None, 'close': 327.3, 'last': 327.3, 'previousClose': None, 'change': None, 'percentage': 0.01019147621988882, 'average': None, 'baseVolume': None, 'quoteVolume': 69.71557, 'info': {'ask': 327.1, 'baseCurrency': 'BCHBEAR', 'bid': 324.0, 'change1h': 0.033491311216429696, 'change24h': 0.01019147621988882, 'changeBod': 0.027001569858712715, 'enabled': True, 'last': 327.3, 'name': 'BCHBEAR/USD', 'price': 327.1, 'priceIncrement': 0.1, 'quoteCurrency': 'USD', 'quoteVolume24h': 69.71557, 'sizeIncrement': 0.0001, 'type': 'spot', 'underlying': None, 'volumeUsd24h': 69.71557}}


{'bids': [[324.0, 15.9222], [322.0, 49.9221], [320.8, 58.1263], [319.3, 136.0149], [317.5, 147.4773], [316.5, 15.8836], [314.4, 201.8833], [314.1, 168.5133], [312.9, 253.0786], [312.6, 93.424]], 'asks': [[327.1, 48.4857], [327.9, 156.5834], [328.4, 17.1807], [330.6, 221.2449], [330.7, 105.9204], [332.2, 242.9397], [332.9, 206.9859], [334.2, 264.8075], [340.0, 310.796], [340.6, 273.1014]], 'timestamp': None, 'datetime': None, 'nonce': None}
```

**Comment by xmatthias** [comment]
The logs don't show any problem. 
You have ~20 open trades - and ` Wallets synced` is called once per pair.

The Pid spamming has been reduced some time ago, and is only shown once every 60 seconds.
In the logs, it is there if you search for `Bot heartbeat` (happened at `2020-02-02 00:33:13,439`).
since the logs only include a little more than 1minute it's perfectly fine to only see that once.

So again - i'm not sure i see any problem here ... the logs at least don't indicate one.

**Comment by Dante2333** [comment]
I think i found an issue. I switched to Dry-run and check this out:

```
2020-02-02 10:33:31,261 - freqtrade.wallets - INFO - Wallets synced.
2020-02-02 10:33:31,268 - freqtrade.wallets - INFO - Wallets synced.
2020-02-02 10:33:31,318 - freqtrade.wallets - INFO - Wallets synced.
2020-02-02 10:33:31,320 - freqtrade.freqtradebot - INFO - Buy signal found: about create a new trade with stake_amount: 60 ...
2020-02-02 10:33:31,320 - freqtrade.freqtradebot - INFO - Using Last Ask / Last Price
2020-02-02 10:33:31,354 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': buy, 'exchange': 'Ftx', 'pair': 'TOMOBEAR/USD', 'limit': 37.38, 'order_type': 'market', 'stake_amount': 59.9949, 'stake_currency': 'USD', 'fiat_currency': 'USD'}
2020-02-02 10:33:31,727 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=None, pair=TOMOBEAR/USD, amount=1.60500000, open_rate=37.38000000, open_since=closed)
2020-02-02 10:33:31,728 - freqtrade.freqtradebot - INFO - Applying fee on amount for Trade(id=None, pair=TOMOBEAR/USD, amount=1.60500000, open_rate=37.38000000, open_since=closed) failed: myTrade-Dict empty found
2020-02-02 10:33:31,728 - freqtrade.persistence - INFO - Updating trade (id=None) ...
2020-02-02 10:33:31,728 - freqtrade.persistence - INFO - MARKET_BUY has been fulfilled for Trade(id=None, pair=TOMOBEAR/USD, amount=1.60500000, open_rate=37.38000000, open_since=closed).
2020-02-02 10:33:31,733 - freqtrade.wallets - INFO - Wallets synced.
2020-02-02 10:33:31,740 - freqtrade.wallets - INFO - Wallets synced.
2020-02-02 10:33:31,777 - freqtrade.wallets - INFO - Wallets synced.
2020-02-02 10:33:31,779 - freqtrade.freqtradebot - INFO - Buy signal found: about create a new trade with stake_amount: 60 ...
2020-02-02 10:33:31,779 - freqtrade.freqtradebot - INFO - Using Last Ask / Last Price
2020-02-02 10:33:31,817 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': buy, 'exchange': 'Ftx', 'pair': 'TRXBULL/USD', 'limit': 42.63, 'order_type': 'market', 'stake_amount': 59.980410000000006, 'stake_currency': 'USD', 'fiat_currency': 'USD'}
2020-02-02 10:33:32,195 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=None, pair=TRXBULL/USD, amount=1.40700000, open_rate=42.63000000, open_since=closed)
2020-02-02 10:33:32,195 - freqtrade.freqtradebot - INFO - Applying fee on amount for Trade(id=None, pair=TRXBULL/USD, amount=1.40700000, open_rate=42.63000000, open_since=closed) failed: myTrade-Dict empty found
2020-02-02 10:33:32,195 - freqtrade.persistence - INFO - Updating trade (id=None) ...
2020-02-02 10:33:32,195 - freqtrade.persistence - INFO - MARKET_BUY has been fulfilled for Trade(id=None, pair=TRXBULL/USD, amount=1.40700000, open_rate=42.63000000, open_since=closed).
2020-02-02 10:33:32,201 - freqtrade.wallets - INFO - Wallets synced.
```

why does it keep mentioning id=None?

**Comment by xmatthias** [comment]
you're using market orders to buy. 

These fill immediately - so the output is generated before the trade is persisted (saved to the database).
It'll be assinged an ID at that point (you'll see that these trades will have the next possible id in the database).

**Comment by hroff-1902** [comment]
Maybe add a small note into faq.md about this (if you use market orders, and see `id=None` in the log, that's okay...)

**Comment by Dante2333** [comment]
This is odd, it seems that proper current rate gets updated correctly on dry-run mode.

**Comment by xmatthias** [comment]
well honestly, it's not so much about market orders - the same can happen without dry-mode and limit orders - if the limit-order fills immediately.

**Comment by xmatthias** [comment]
I think the behaviour can best be replicated by using one single pair.

mine was BTC/USD on FTX.

Notice the change in current_rate from 3989.5 to 3989.0 after calling /status a few times.
This was also aligned to the exchange price.

![2020-02-03-070211_547x680_scrot](https://user-images.githubusercontent.com/5024695/73629534-c932c900-4653-11ea-9e30-27746be1dea7.png)

Checking back on the trade a bit later shows that the price moved again - this time to 9381.5.

![2020-02-03-070839_560x681_scrot](https://user-images.githubusercontent.com/5024695/73629617-04cd9300-4654-11ea-9d01-b28b8429ff23.png)

**Comment by Dante2333** [comment]
Dry-run mode is working. I tried setting up a new server and fresh start using the proper virtual environment as instructed per the installation tutorial on freqtrade.io and to no avail I cant get current rates to work properly under real trading.

**Comment by xmatthias** [comment]
Please ensure that all configuration settings are identical between the dry and test instance, exept for "key", "secret" and "dry_run". 
Use the same strategy you used on dry-run (not that that would matter how prices are determined).

What you can try is leave dry_run to on - but provide valid keys.
That will log you into the exchange - but freqtrade won't place orders.

Now if what i suspect is true (different results depending on the login method) this may have the same effect than if you're running live.
On the other hand - another suspicion is that the results depend if you've got a position in that asset or not - which would not be confirmed by this.


If that's the case, then the only way for us to help you is by adding FTX to the "blocked" list of exchanges.

**Comment by Dante2333** [comment]
Ok, will report back in 48 hrs

**Comment by Dante2333** [comment]
I've started my 2 ftx bots, 1 dry run 1 with real exchange, will report back.

**Comment by Dante2333** [comment]
![1](https://user-images.githubusercontent.com/5402278/74329712-b19bc480-4d66-11ea-8702-4cf8a24d61e1.jpg)
![2](https://user-images.githubusercontent.com/5402278/74329713-b19bc480-4d66-11ea-89ff-6a4fa9ccca0a.jpg)
[1log.txt](https://github.com/freqtrade/freqtrade/files/4191662/1log.txt)

**Comment by xmatthias** [comment]
There seems to be one logfile attached (1log.txt) - which bot is that for?

**Comment by Dante2333** [comment]
> There seems to be one logfile attached (1log.txt) - which bot is that for?

The one running with real funds. I could attach the other one for dry run if you want.

**Comment by xmatthias** [comment]
one more question ... what's the timedifference to take into account? the logs are from 10:33 - 10:59 ... 
are you (or the server) in a timezone -4 - so the the timestamp of 6:54 aligns with the timestamp at 10:54?

**Comment by xmatthias** [comment]
based on "very quickly" looking through the logs - it seems like freqtrade assumes the orders to still be open / on the exchange.

did they fill on the exchange?
if yes - please run the following for me (requires your exchange keys - so it's not for me to run this ...)

in the database (pick a random order id)
``` sql
select id, pair, open_order_id from trades where id = 1;
```

take note of the "open_order_id" field and use it in the below query:

``` python
import ccxt
ct = ccxt.ftx({'apiKey': '<yourAPIKey>', 'secret': '<your_secret'})
order = ct.fetch_order('<order_id_from_database>', 'LSK/BTC')
print(order)
```

**Comment by Dante2333** [comment]
```{'info': {'avgFillPrice': 2666.0, 'clientId': None, 'createdAt': '2020-02-12T00:53:49.009726+00:00', 'filledSize': 0.0007, 'future': None, 'id': 3109208514, 'ioc': True, 'market': 'BNBBULL/USD', 'postOnly': False, 'price': None, 'reduceOnly': False, 'remainingSize': 0.0, 'side': 'buy', 'size': 0.0007, 'status': 'closed', 'type': 'market'}, 'id': '3109208514', 'timestamp': 1581468829009, 'datetime': '2020-02-12T00:53:49.009Z', 'lastTradeTimestamp': None, 'symbol': 'BNBBULL/USD', 'type': 'market', 'side': 'buy', 'price': None, 'amount': 0.0007, 'cost': 4.9e-07, 'average': None, 'filled': 0.0007, 'remaining': 0.0, 'status': 'closed', 'fee': None, 'trades': None}
```

**Comment by xmatthias** [comment]
Can you please run the same again for the same order, but this time in verbose mode:

```
import ccxt
ct = ccxt.ftx({'apiKey': '<yourAPIKey>', 'secret': '<your_secret'})
ct.verbose=True
order = ct.fetch_order('3109208514', 'BNBBULL/USD')
print(order)
```

This should print a ton of info - i'll need most of it (however it may include your API keys too - so please search for these before posting - and/or send it via slack PM - still after removing the obvious API Keys).
I don't necessarily need the order output itself (that should be aligned to the above) - but this will also include the raw queries done against the exchange.

I won't do anything with them - but i don't even want to see them and if i do, i'll tell you to revoke the API Keys

**Comment by xmatthias** [comment]
I think i've found the bug - ccxt is not returning price - which is one way for us to detect that an order is closed.
Because of that - freqtrade assumes that the order is still open - and treats it as such.

While looking at it i found another issue - both are detailed in the ccxt issue
https://github.com/ccxt/ccxt/issues/6479.

**Comment by Dante2333** [comment]
Well. Fresh in the oven

After updating to latest ccxt, now we get 
 
`Traceback (most recent call last):
  File "/home/dante/freqtrade/freqtrade/main.py", line 36, in main
    return_code = args['func'](args)
  File "/home/dante/freqtrade/freqtrade/commands/trade_commands.py", line 20, in start_trading
    worker.run()
  File "/home/dante/freqtrade/freqtrade/worker.py", line 62, in run
    state = self._worker(old_state=state)
  File "/home/dante/freqtrade/freqtrade/worker.py", line 98, in _worker
    self._throttle(func=self._process, min_secs=throttle_secs)
  File "/home/dante/freqtrade/freqtrade/worker.py", line 111, in _throttle
    result = func(*args, **kwargs)
  File "/home/dante/freqtrade/freqtrade/worker.py", line 121, in _process
    self.freqtrade.process()
  File "/home/dante/freqtrade/freqtrade/freqtradebot.py", line 156, in process
    self.enter_positions()
  File "/home/dante/freqtrade/freqtrade/freqtradebot.py", line 227, in enter_positions
    trades_created += self.create_trade(pair)
  File "/home/dante/freqtrade/freqtrade/freqtradebot.py", line 421, in create_trade
    return self.execute_buy(pair, stake_amount)
  File "/home/dante/freqtrade/freqtrade/freqtradebot.py", line 475, in execute_buy
    time_in_force=time_in_force)
  File "/home/dante/freqtrade/freqtrade/exchange/exchange.py", line 507, in buy
    return self.create_order(pair, ordertype, 'buy', amount, rate, params)
  File "/home/dante/freqtrade/freqtrade/exchange/exchange.py", line 478, in create_order
    amount, rate_for_order, params)
  File "/home/dante/.local/lib/python3.7/site-packages/ccxt/ftx.py", line 917, in create_order
    return self.parse_order(result, market)
  File "/home/dante/.local/lib/python3.7/site-packages/ccxt/ftx.py", line 811, in parse_order
    cost = filled * price
TypeError: unsupported operand type(s) for *: 'float' and 'NoneType'
 `

**Comment by xmatthias** [comment]
Would you mind running the query from above again with the new version ... and then post the outcome to the ccxt issue?

```
import ccxt
ct = ccxt.ftx({'apiKey': '<yourAPIKey>', 'secret': '<your_secret'})
# ct.verbose=True   # Not needed here i think
order = ct.fetch_order('3109208514', 'BNBBULL/USD')
print(order)
```

**Comment by Dante2333** [comment]
`{'info': {'avgFillPrice': 2666.0, 'clientId': None, 'createdAt': '2020-02-12T00:53:49.009726+00:00', 'filledSize': 0.0007, 'future': None, 'id': 3109208514, 'ioc': True, 'market': 'BNBBULL/USD', 'postOnly': False, 'price': None, 'reduceOnly': False, 'remainingSize': 0.0, 'side': 'buy', 'size': 0.0007, 'status': 'closed', 'type': 'market'}, 'id': '3109208514', 'timestamp': 1581468829009, 'datetime': '2020-02-12T00:53:49.009Z', 'lastTradeTimestamp': None, 'symbol': 'BNBBULL/USD', 'type': 'market', 'side': 'buy', 'price': 2666.0, 'amount': 0.0007, 'cost': 1.8662, 'average': 2666.0, 'filled': 0.0007, 'remaining': 0.0, 'status': 'closed', 'fee': None, 'trades': None}
`

**Comment by xmatthias** [comment]
ah - my bad - the problem seems to come from a different method now....
i'll update the ccxt issue with that

**Comment by xmatthias** [comment]
@Dante2333 can you check with ccxt version 1.22.45? (run `pip install -U ccxt` ...)

thanks in advance

**Comment by Dante2333** [comment]
ccxt 1.22.46 `{'info': {'avgFillPrice': 2666.0, 'clientId': None, 'createdAt': '2020-02-12T00:53:49.009726+00:00', 'filledSize': 0.0007, 'future': None, 'id': 3109208514, 'ioc': True, 'market': 'BNBBULL/USD', 'postOnly': False, 'price': None, 'reduceOnly': False, 'remainingSize': 0.0, 'side': 'buy', 'size': 0.0007, 'status': 'closed', 'type': 'market'}, 'id': '3109208514', 'timestamp': 1581468829009, 'datetime': '2020-02-12T00:53:49.009Z', 'lastTradeTimestamp': None, 'symbol': 'BNBBULL/USD', 'type': 'market', 'side': 'buy', 'price': 2666.0, 'amount': 0.0007, 'cost': 1.8662, 'average': 2666.0, 'filled': 0.0007, 'remaining': 0.0, 'status': 'closed', 'fee': None, 'trades': None}
`

**Comment by Dante2333** [comment]
It works perfectly now, the balance is updated.
